### PR TITLE
INGM-289: add mypy

### PR DESCRIPTION
**Pull request by WoodElv**
### Type of change

Please add a description and delete options that are not relevant.

- [x] Add mypy into Pipfile.
- [x] Add typing for all functions of the library.
- [x] Check there is no dynamic typing.
- [x] Add static analysis step for Jenkins process.

### Tests
- [ ] Add new unit tests if it applies.
- [x] Run tests.

Please describe the tests that you ran to verify your changes if it applies. 

### Documentation

Please update the documentation.

- [x] Update docstrings of every function, method or class that change.
- [x] Use [type hints](https://docs.python.org/3/library/typing.html) for every function and argument.
- [x] Build documentation locally to verify changes.
- [x] Add the changes at the `[Unreleased]` section of the [CHANGELOG](CHANGELOG.md).

### Code formatting

- [x] Use black package to format the code: `black -l 100 ingeniamotion tests`. It is recommended to configure the code editor to automatically format the code using black with a max length line of 100.

### CODE REVIEW

File: ingeniamotion/communication.py [hunk 1]
```diff
@@ -282,7 +288,7 @@ def __get_adapter_name(address: str) -> Optional[str]:
         for adapter in ifaddr.get_adapters():
             for ip in adapter.ips:
                 if ip.is_IPv4 and ip.ip == address:
-                    return adapter.name.decode("utf-8")
+                    return f"{adapter.name.decode('utf-8')}"
```
[the remark this answers is not in the document]
Comment (MartinALL91) [reply ingeniamotion/communication.py:293]: You could do: `return bytes.decode(adapter.name)`

File: ingeniamotion/configuration.py [hunk 2]
```diff
@@ -533,9 +553,12 @@ def get_motor_pair_poles(self, servo: str = DEFAULT_SERVO, axis: int = DEFAULT_A
             Pair poles value.
 
         """
-        return self.mc.communication.get_register(
+        pair_poles = self.mc.communication.get_register(
             self.MOTOR_POLE_PAIRS_REGISTER, servo=servo, axis=axis
         )
+        if not isinstance(pair_poles, int):
+            raise IMException("Pair poles value has to be an integer")
```
Comment (MartinALL91) [inline ingeniamotion/configuration.py:560]: I think the message should be: 
"Received wrong value type when reading {self.MOTOR_POLE_PAIRS_REGISTER}. Got type {type(pair_poles)}, expected int type"

**Review by MartinALL91 — CHANGES_REQUESTED** [review]

File: ingeniamotion/capture.py [hunk 3]
```diff
@@ -364,8 +370,8 @@ def disable_monitoring(
             return
         drive = self.mc.servos[servo]
         drive.monitoring_disable()
-        if version >= MonitoringVersion.MONITORING_V3:
-            return drive.monitoring_remove_data()
```
[the remark this answers is not in the document]
Comment (WoodElv) [reply ingeniamotion/capture.py:368]: You are right!

File: ingeniamotion/capture.py [hunk 4]
```diff
@@ -386,7 +392,8 @@ def disable_disturbance(
             return self.disable_monitoring(servo=servo, version=version)
         drive = self.mc.servos[servo]
         drive.disturbance_disable()
-        return drive.disturbance_remove_data()
```
[the remark this answers is not in the document]
Comment (WoodElv) [reply ingeniamotion/capture.py:389]: Done.

File: ingeniamotion/capture.py [hunk 5]
```diff
@@ -107,12 +111,18 @@ def create_poller(
 
         Raises:
             IMRegisterNotExist: If register does not exist in dictionary.
+            TypeError: If some parameter has an error type.
```
Comment (GullinBustin) [inline ingeniamotion/capture.py:114]: error type? I think it should be "wrong type". "If some parameter has a wrong type"

Comment (GullinBustin) [reply ingeniamotion/capture.py:114]: That comment affects a lot of lines

File: ingeniamotion/capture.py [hunk 6]
```diff
@@ -415,10 +436,15 @@ def get_monitoring_status(self, servo: str = DEFAULT_SERVO) -> int:
 
         Raises:
             IMRegisterNotExist: If the register doesn't exist.
+            TypeError: If some parameter has an error type.
```
Comment (GullinBustin) [inline ingeniamotion/capture.py:439]: Technically is not a parameter, it is the read value type.

Comment (GullinBustin) [reply ingeniamotion/capture.py:439]: That comment affects a lot of lines

File: ingeniamotion/comkit.py [hunk 7]
```diff
@@ -157,7 +163,7 @@ def set_attributes(src_tree, dest_tree) -> Element:
     return dest_tree
 
 
-def create_attribute(attribute, src_root, dest_root, core) -> None:
+def create_attribute(attribute: str, src_root: Element, dest_root: Element, core: CORE) -> None:
```
Comment (GullinBustin) [inline ingeniamotion/comkit.py:166]: Docstrings says this function returns "Destination tree", but it returns None

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py [hunk 8]
```diff
@@ -17,17 +22,19 @@ class DigitalHallTest(Feedbacks):
...
     @BaseTest.stoppable
-    def feedback_setting(self):
+    def feedback_setting(self) -> None:
         self.halls_extra_settings()
         super().feedback_setting()
 
     @BaseTest.stoppable
-    def halls_extra_settings(self):
+    def halls_extra_settings(self) -> None:
+        if not isinstance(self.pair_poles, int):
```
Comment (GullinBustin) [inline ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py:36]: Probably should be more accurate check  if pair_poles is not None, it is already defined as Optional[int]

File: ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py [hunk 9]
```diff
@@ -53,8 +60,10 @@ def halls_extra_settings(self):
             del self.backup_registers[self.axis][self.VELOCITY_FEEDBACK_FILTER_1_FREQUENCY_REGISTER]
 
     @BaseTest.stoppable
-    def suggest_polarity(self, pol):
+    def suggest_polarity(self, pol: Feedbacks.Polarity) -> None:
         polarity_uid = self.FEEDBACK_POLARITY_REGISTER
         pair_poles_uid = self.DIG_HALL_POLE_PAIRS_REGISTER
+        if not isinstance(self.pair_poles, int):
```
Comment (GullinBustin) [inline ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py:66]: Same here, check if it is None is more accurate

File: ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py [hunk 10]
```diff
@@ -75,36 +78,40 @@ class Polarity(IntEnum):
         "CL_VEL_FBK_FILTER1_FREQ",
     ]
 
-    FEEDBACK_POLARITY_REGISTER = None
+    FEEDBACK_POLARITY_REGISTER = ""
```
Comment (GullinBustin) [inline ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py:81]: You can write `FEEDBACK_POLARITY_REGISTER: str`

File: ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py [hunk 11]
```diff
@@ -359,24 +388,28 @@ def loop(self):
...
+    ) -> ResultType:
         test_output = 0
 
         test_output += self.check_symmetry(position_displacement, negative_displacement)
         test_output += self.check_resolution(position_displacement)
         test_output = self.ResultType.SUCCESS if test_output == 0 else test_output
         polarity = self.check_polarity(position_displacement)
         self.suggest_polarity(polarity)
-        return test_output
+        return self.ResultType(test_output)  # TODO check return
```
Comment (GullinBustin) [inline ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py:401]: I don't understand the comment

File: ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py [hunk 12]
```diff
@@ -359,24 +388,28 @@ def loop(self):
...
 
-    def get_result_msg(self, output):
+    def get_result_msg(self, output: ResultType) -> str:
         if output == self.ResultType.SUCCESS:
             return self.result_description[output]
         if output < 0:
             text = [self.result_description[x] for x in self.result_description if -output & -x > 0]
             return ".".join(text)
+        else:
+            raise IMException("Unknown error")
```
Comment (GullinBustin) [inline ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py:410]: what?

File: ingeniamotion/wizard_tests/stoppable.py [hunk 13]
```diff
@@ -1,35 +1,42 @@
+import typing
 from time import time
 from functools import wraps
+from typing import TypeVar, Callable, Any
+
+F = TypeVar("F", bound=Callable[..., Any])
```
Comment (GullinBustin) [inline ingeniamotion/wizard_tests/stoppable.py:6]: This is not in use

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: ingeniamotion/communication.py [hunk 14]
```diff
@@ -282,7 +290,7 @@ def __get_adapter_name(address: str) -> Optional[str]:
         for adapter in ifaddr.get_adapters():
             for ip in adapter.ips:
                 if ip.is_IPv4 and ip.ip == address:
-                    return adapter.name.decode("utf-8")
+                    return bytes.decode(adapter.name)
```
[the remark this answers is not in the document]
Comment (MartinALL91) [reply ingeniamotion/communication.py:293]: It applies "utf-8" encodign by default https://docs.python.org/3/library/stdtypes.html#bytes.decode

[the remark this answers is not in the document]
Comment (WoodElv) [reply ingeniamotion/communication.py:293]: Exactly.

File: ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py (hunk 8 above)
Comment (WoodElv) [reply ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py:36]: I think it's already typed well. If pair_poles variable was a None, the register couldn't be set correctly. In addition, the set_register function doesn't accept None type. Maybe the best solution would be typing pair_poles to int instead. What do you think?

File: ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py (hunk 9 above)
Comment (WoodElv) [reply ingeniamotion/wizard_tests/feedbacks_tests/digital_hall_test.py:66]: Same as above.

File: ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py [hunk 15]
```diff
@@ -114,14 +121,18 @@ def check_symmetry(self, positive, negative):
...
         self.logger.info("Feedback polarity detected: %s", polarity.name)
         return polarity
 
     @BaseTest.stoppable
-    def check_resolution(self, displacement):
+    def check_resolution(self, displacement: float) -> ResultType:
+        if not isinstance(self.pair_poles, int):
+            raise TypeError("Pair poles has to be set before resolution checking.")
+        if not isinstance(self.feedback_resolution, int):
+            raise TypeError("Feedbacks has to be set before resolution checking.")
```
Comment (GullinBustin) [inline ingeniamotion/wizard_tests/feedbacks_tests/feedback_test.py:135]: `Feedback resolution has to be set before resolution checking.`
In addition, the if condition for pair_poles and feedback_resolution should be " is None"

**Review by GullinBustin — CHANGES_REQUESTED** [review]

**Review by MartinALL91 — APPROVED** [review]

**Review by GullinBustin — APPROVED** [review]

**Pull request merged**